pandas/core/groupby/ops.py

```diff
@@ -726,10 +726,15 @@ def _get_group_keys(self):
             return get_flattened_list(ids, ngroups, self.levels, self.codes)
 
     @final
-    def apply(self, f: F, data: FrameOrSeries, axis: int = 0):
+    @cache_readonly
+    def group_keys_seq(self):
```

Comment:
why are you adding this?

Reply:
without adding this, the PR makes us call _get_group_keys repeatedly; this is a cached version of that call

Reply:
updated to just remove _get_group_keys and only have the cached version